pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/JavaVisitorBase.java

```diff
@@ -0,0 +1,375 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.java.ast;
+
+import net.sourceforge.pmd.lang.ast.AstVisitorBase;
+
+/**
+ * Base implementation of {@link JavaVisitor}. This adds delegation logic
+ * which the interface doesn't have.
+ *
+ * <p>Contrary to the old visitor, which used Object as both parameter and
+ * return type, this visitor uses separate type parameters for those. This
+ * means you can't just return the parameter, unless your visitor has equal
+ * parameter and return type. This type signature subsumes many possible
+ * signatures. The old one is {@code <Object, Object>}, still implemented
+ * by {@link JavaParserVisitor} for backwards compatibility. If you don't
+ * want to return a value, or don't want a parameter, use {@link Void}.
+ *
+ * <p>Since 7.0.0 we use default methods on the interface, which removes
+ * code duplication. However it's still recommended to extend a base class,
+ * for forward compatibility.
+ */
+public class JavaVisitorBase<P, R> extends AstVisitorBase<P, R> implements JavaVisitor<P, R> {
+
+    // todo on java-grammar: uncomment
+
+
+    // <editor-fold defaultstate="collapsed" desc="Methods/constructors">
```

Comment:
Is that IDEA specific?

Reply:
This works in Netbeans and IntelliJ

Reply:
Would you rather I remove it?

Reply:
Needing to fold could be a sign of the file being too big and indicate that we maybe should split it. But that's just to keep in mind and we'll need to revisit this later.